compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala

```diff
@@ -111,11 +111,23 @@ object ErrorReporting {
       errorTree(tree, typeMismatchMsg(normalize(tree.tpe, pt), pt, implicitFailure.postscript))
 
     /** A subtype log explaining why `found` does not conform to `expected` */
-    def whyNoMatchStr(found: Type, expected: Type) =
-      if (ctx.settings.explaintypes.value)
+    def whyNoMatchStr(found: Type, expected: Type) = {
+      def dropJavaMethod(tp: Type): Type = tp match {
+        case tp: PolyType =>
+          tp.derivedPolyType(resType = dropJavaMethod(tp.resultType))
+        case tp: JavaMethodType =>
+          MethodType(tp.paramNames, tp.paramTypes, dropJavaMethod(tp.resultType))
+        case tp => tp
+      }
+      val found1 = dropJavaMethod(found)
+      val expected1 = dropJavaMethod(expected)
+      if ((found1 eq found) != (expected eq expected1) && (found1 <:< expected1))
+        "\n (Note that Scala's and Java's representation of this type differs)"
```

Comment:
This might not be enough for the user to understand the issue. In the example we would get 
```
overriding method toArray in trait Collection of type [T](x$0: Array[T])Array[T];
       method toArray of type [T](a: Array[T])Array[T] has incompatible type
       (Note that Scala's and Java's representation of this type differs)
```

Would it be possible to hint that to make the representation match one would need to write  
`def toArray[T](a: Array[T with Object]): Array[T with Object]`?